metis/lib/server/controllers/folder_controller.rb

```diff
@@ -23,6 +23,18 @@ def list
     success_json(files: files, folders: folders)
   end
 
+  def list_all_folders
+    bucket = require_bucket
+
+    folders = Metis::Folder.where(
+      bucket: bucket
+    ).all.map do |fold|
+      fold.to_hash
+    end
```

Comment:
It would be pretty easy to do this in a cheaper way that avoids repeated recursive queries to the db. Since in your original query you actually have all of the folder records loaded, along with their folder_id, there is no need to make return queries to the database. Just group_by folder_id, then you should be able to do a traverse of the tree in O(n) and list all of the folder paths in the bucket.

Reply:
:+1:  ^^

Reply:
Also does this potentially need a paging operation?  I can imagine a bucket easily having 10+ items.

Reply:
:+1: Let me test out grouping and see if that helps. Pagination may make sense.